db/migrate/20211105044524_update_patients_latest_assessment_at_by_assessment_reported_at.rb

```diff
@@ -0,0 +1,49 @@
+class UpdatePatientsLatestAssessmentAtByAssessmentReportedAt < ActiveRecord::Migration[6.1]
+  def change
+    reversible do |dir|
+      dir.up { populate_latest_assessment_fields(:reported_at) }
+      dir.down { populate_latest_assessment_fields(:created_at) }
+    end
+  end
+
+  def populate_latest_assessment_fields(field)
+    # reset :latest_assessment_at and :latest_assessment_symptomatic to NULL
+    execute <<-SQL.squish
+      UPDATE patients
+      SET patients.latest_assessment_at = NULL, patients.latest_assessment_symptomatic = NULL
+      WHERE patients.purged = FALSE
+    SQL
+
+    # populate :latest_assessment_at
+    execute <<-SQL.squish
+      UPDATE patients
+      INNER JOIN (
+        SELECT patient_id, MAX(#{field}) AS latest_assessment_at
+        FROM assessments
+        GROUP BY patient_id
+      ) t ON patients.id = t.patient_id
+      SET patients.latest_assessment_at = t.latest_assessment_at
+      WHERE patients.purged = FALSE
```

Comment:
The only time I can think that `latest_assessment_at` would be wrong is in the case that `reported_at` and `created_at` are not equal. Unless we can come up with a case where `reported_at` and `created_at` are equal, and we still need to update `latest_assessment_at`, I think we should add a `WHERE` to scope this query to just the situation where `created_at` and `reported_at` are not equal. Same goes for the query below. In my mind this accomplishes two goals:
1. Speeds up our migration
2. Avoids any potential impact on most Patients. This feature was only added recently, and for all patients who self report (which I believe is most patients), we are not going to have to worry about `reported_at` and `created_at` being out of sync, since a patient cannot set `reported_at` to anything different than created at. This query looks good to me and as far as I can tell it should work with all patients. But I'd feel better if we mitigate the risk by avoiding updating the patients we don't need to update, which should be most patients.

Reply:
that makes sense, done: https://github.com/SaraAlert/SaraAlert/pull/1154/commits/5f4dece3f2e0e271e804c619b4dae3ea2de20592